stablehlo/reference/NumPy.cpp

```diff
@@ -0,0 +1,351 @@
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Invalid NumPy header.");
+
+  header.erase(std::remove_if(header.begin(), header.end(), isspace),
+               header.end());
+
+  auto wordSizeOrError = parseDescrHeader<T>(header);
+  if (!wordSizeOrError)
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Failed to parse descr header.");
+
+  if (parseFortranOrderKey(header))
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Failed to parse fortran_order header.");
+
+  auto shapeOrError = parseShapeHeader(header);
+  if (!shapeOrError)
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Failed to parse shape header.");
+
+  shape = *shapeOrError;
+  wordSize = *wordSizeOrError;
+
+  return llvm::Error::success();
+}
+
+namespace {
+template <typename T>
+class ToNumpy {
+ public:
+  llvm::Error operator()(StringRef filename, ShapedType type,
+                         const char* data) {
+    int fd;
+    if (llvm::sys::fs::openFileForWrite(filename, fd,
+                                        llvm::sys::fs::CD_CreateAlways))
+      return llvm::createStringError(llvm::errc::io_error,
+                                     "Failed to open NumPy file.");
+
+    llvm::raw_fd_ostream out(fd, /*shouldClose=*/true);
+
+    buildNumpyHeader<T>(out, type.getShape());
+    out.write(data, sizeof(T) * type.getNumElements());
+
+    return llvm::Error::success();
+  }
+};
+
+template <typename T>
+class FromNumpy {
+ public:
+  llvm::ErrorOr<Tensor> operator()(StringRef filename, ShapedType type) {
+    std::ifstream in(filename.str(), std::ifstream::binary);
+    size_t wordSize = 0;
+
+    if (readNumpyHeader<T>(in, wordSize, type.getShape()))
+      return llvm::errc::invalid_argument;
+
+    const int64_t numLoadedElements = std::accumulate(
+        type.getShape().begin(), type.getShape().end(), 1, std::multiplies<>());
+
+    if (numLoadedElements != type.getNumElements() || wordSize != sizeof(T))
+      return llvm::errc::invalid_argument;
+
+    const size_t dataBytesToRead = wordSize * numLoadedElements;
+    std::vector<T> data(dataBytesToRead);
+    in.read(reinterpret_cast<char*>(data.data()), dataBytesToRead);
+
+    return Tensor(type,
+                  HeapAsmResourceBlob::allocateAndCopyInferAlign<T>(data));
+  }
+};
+
+template <template <typename Type> class Functor, typename... Args>
+static decltype(auto) dispatchType(Type type, Args&&... args) {
+  if (auto quantized = type.dyn_cast<quant::QuantizedType>()) {
+    switch (quantized.getStorageTypeIntegralWidth()) {
+      case 8:
+        return Functor<int8_t>()(std::forward<Args>(args)...);
+      case 16:
+        return Functor<int16_t>()(std::forward<Args>(args)...);
+      default:
+        llvm::report_fatal_error("Unhandled quantized type");
+    }
+  }
```

Comment:
Hm, is it worthwhile to add this back in if/when it is supported? Or should we leave it in so it works OOB once support is added? (deferring to @ghpvnist )

Reply:
I'd prefer removal. It's best to consciously add support for these ops instead of predict how the undesigned interpreter feature will work in the future. I.e. maybe we need `QuantizedTensor` type and very specific ops support this new type.. now this code will never be hit

Reply:
Sounds good, I've removed this from this PR.

Reply:
I also prefer removing unused features until it is fully supported. Things may change which could make this change stale or break.